DeepSpeech.py

```diff
@@ -534,68 +498,66 @@ def run_set(set_name):
-                    train_loss = run_set('train')
+                    train_loss = run_set('train', train_init_op, train_batches)
                     log_info('Finished training epoch %d - loss: %f' % (current_epoch, train_loss))
                     checkpoint_saver.save(session, checkpoint_path, global_step=global_step)
-                    steps_trained = 0
-                    # Validation
-                    log_info('Validating epoch %d ...' % current_epoch)
-                    dev_loss = run_set('dev')
-                    dev_losses.append(dev_loss)
-                    log_info('Finished validating epoch %d - loss: %f' % (current_epoch, dev_loss))
-                    if dev_loss < best_dev_loss:
-                        best_dev_loss = dev_loss
-                        save_path = best_dev_saver.save(session, best_dev_path, latest_filename=best_dev_filename)
-                        log_info("Saved new best validating model with loss %f to: %s" % (best_dev_loss, save_path))
-                    # Early stopping
-                    if FLAGS.early_stop and len(dev_losses) >= FLAGS.es_steps:
-                        mean_loss = np.mean(dev_losses[-FLAGS.es_steps:-1])
-                        std_loss = np.std(dev_losses[-FLAGS.es_steps:-1])
-                        dev_losses = dev_losses[-FLAGS.es_steps:]
-                        log_debug('Checking for early stopping (last %d steps) validation loss: '
-                                  '%f, with standard deviation: %f and mean: %f' %
-                                  (FLAGS.es_steps, dev_losses[-1], std_loss, mean_loss))
-                        if dev_losses[-1] > np.max(dev_losses[:-1]) or \
-                           (abs(dev_losses[-1] - mean_loss) < FLAGS.es_mean_th and std_loss < FLAGS.es_std_th):
-                            log_info('Early stop triggered as (for last %d steps) validation loss:'
-                                     ' %f with standard deviation: %f and mean: %f' %
-                                     (FLAGS.es_steps, dev_losses[-1], std_loss, mean_loss))
-                            break
-                log_debug('Closing queues...')
+
+                    if FLAGS.dev:
+                        # Validation
+                        log_info('Validating epoch %d ...' % current_epoch)
+                        dev_loss = run_set('dev', dev_init_op, dev_batches)
+                        dev_losses.append(dev_loss)
+                        log_info('Finished validating epoch %d - loss: %f' % (current_epoch, dev_loss))
+
+                        if dev_loss < best_dev_loss:
+                            best_dev_loss = dev_loss
+                            save_path = best_dev_saver.save(session, best_dev_path, latest_filename=best_dev_filename)
+                            log_info("Saved new best validating model with loss %f to: %s" % (best_dev_loss, save_path))
+
+                        # Early stopping
```

Comment:
Adding it here, as there is no better place: You should add some wording to the early stopping flags about the requirement of having validation activated (as you made it optional).

Reply:
The documentation for the early_stop flag already mentions it's tied to validation, but I'll try to clarify it.